### PR TITLE
Provide embedded http server

### PR DESCRIPTION
**Pull request by oncletom**
_live from https://www.pagedmedia.org/pagedmedia-prepostprint-workshop/_

- port is selected randomly
- it removes the manual task of starting the server
- it prevents the "fetch() cannot run local files" error

### CODE REVIEW

**Comment by Mogztter** [comment]
Thanks @oncletom 
For reference, here's the reply from the Paged Media about replacing `fetch` with `XMLHttpRequest`: https://gitlab.pagedmedia.org/tools/pagedjs/issues/86#note_480


> Thats interesting that it works different then fetch, I'll try it out and make sure everything works ok with the switch.
However I'd recommend just running an express server for the files rather then allowing access to the whole filesystem like we do in pagedjs-cli: https://gitlab.pagedmedia.org/tools/pagedjs-cli/blob/master/index.js#L48

So in the meantime using an embedded server is the way to go.
Any objection @matklad  @lordofthejars  @mojavelinux ?

Having said that we could revisit this issue when/if Paged.js switch to `XMLHttpRequest` where we could choose to lower the security of the Chrome headless instance (instead of using an embedded server).

**Comment by oncletom** [comment]
Ah sure! 

> Perhaps there should be a flag to disable/enable this behavior?

Yes. I'd eventually couple it with a `--base-url` so as `localhost:5000` is not hardcoded, and can be changed to whatever is the user local setup.

> Would it be possible to make it work with the --watch option <mark>as expected</mark>?

What do you mean? To start the server only with `--watch`?

Let me know for both cases, I'm happy to make the changes.

**Comment by matklad** [comment]
Oh, sorry, looks like I completely misunderstood what this is doing =/

I though this was not for generating PDF which refer to local resources, but for a more streamlined dev workflow. That is, I thought that this opens a browser windows with HTML for the purposes of displaying a rendered document. 

So, my comment about `--watch` was for a completely orthogonal issue. Currently, to get live-reload I run `/asciidoctorjs-pdf` in one terminal and `reload -b` in another (where `reload` is some npm module for live reloading). I with that this workflow was built-in. Though, this is not a big deal, b/c I can create it locally.

**Comment by oncletom** [comment]
Understood.

This PR is solely to make the generation of the PDF possible without launching an HTTP server aside. This HTTP server is not meant to be user facing, but accessible to the headless browser to create a safe environment for the paged.js library.

At the moment, if you don't launch an HTTP server on `localhost:5000`, the PDF generation will fail.

**Comment by Mogztter** [comment]
> At the moment, if you don't launch an HTTP server on localhost:5000, the PDF generation will fail.

on the `feat/pagejs` branch :wink: 

> I opted for a lightweight set of modules. The server is closed once the PDF generation is done.

:+1: 

> Would it be possible to make it work with the --watch option as expected?

I think we should start the HTTP server only once if we use `--watch`. If I'm not mistaken, currently the HTTP server will be opened and closed for each generation ? What do you think ?

**Comment by oncletom** [comment]
Yeah right, it's only on the Paged.js branch 😅

Server is pretty fast to start and close. I'm happy to do the change but not sure if that's necessary to do it now. 

I added a new thing in afa1590, the ability to proxy 3 files (`asciidoctor.css`, `document.css` and `pagedjs.polyfill.js` to the package files). This way the assets are always loaded relatively to the original source document.

I did that because I had an issue while generating a PDF document from an Asciidoc files located outside of the project file structure.

File: lib/converter.js [hunk 1]
```diff
@@ -1,10 +1,28 @@
...
-const fs = require('fs')
+const {writeFile, readFile} = require('fs').promises
 const path = require('path')
-const writeFile = util.promisify(fs.writeFile)
 const puppeteer = require('puppeteer')
 const asciidoctor = require('asciidoctor.js')()
+const getPort = require('get-port')
+const handler = require('serve-handler')
+const {createServer} = require('http')
+const debug = util.debuglog('asciidoctor-pdf')
```
Comment (oncletom) [inline lib/converter.js]: This way, debug logs appear when doing `NODE_DEBUG=asciidoctor-pdf ./bin/asciidoctorjs-pdf ...`

File: lib/converter.js [hunk 2]
```diff
@@ -45,51 +63,102 @@ async function convert(inputFile, options, timings) {
...
+  })
+
+  const [, port] = await Promise.all([
+    writeFile(tempFile, html),
+    getPort(),
+  ])
+
+  const url = `http://localhost:${port}/${tempFilename}`;
+  const httpOptions = {
+    public: workingDir,
```
Comment (oncletom) [inline lib/converter.js]: With this, public assets are relative to the original document. It prevents the need for `relativeFilePath` and `baseAbsolutePath `.

File: lib/converter.js [hunk 3]
```diff
@@ -45,51 +63,102 @@ async function convert(inputFile, options, timings) {
...
+    cleanUrls: false,
+  };
+
+  return await new Promise((resolve, reject) => {
+    debug(`start server on localhost:${port}`)
+
+    const server = createServer(async (req, res) => {
+      if (req.url in STATIC_ASSETS) {
```
Comment (oncletom) [inline lib/converter.js]: I was confused by `serve-handler`'s `rewrites` options. It looks like this works only with files located within its public directory.

File: lib/converter.js [hunk 4]
```diff
@@ -45,51 +63,102 @@ async function convert(inputFile, options, timings) {
...
+        return handler(req, res, httpOptions)
+      }
+    })
+      .listen(port, async () => {
+        browser.on('error', reject)
+        server.on('error', reject)
+        page.on('error', reject)
+
+        try {
+          debug(`goto ${url}`)
```
Comment (oncletom) [inline lib/converter.js]: I encountered timeouts with large HTML pages. Better to make the server crash.

File: lib/converter.js [hunk 5]
```diff
@@ -1,10 +1,28 @@
...
+const handler = require('serve-handler')
+const {createServer} = require('http')
+const debug = util.debuglog('asciidoctor-pdf')
+
+const STATIC_ASSETS = {
+  '/asciidoctor.css': {
+    src: path.join(__dirname, '..', 'examples', 'document', 'asciidoctor.css'),
+    type: 'text/css',
+  },
+  '/asciidoctor-document.css': {
```
Comment (Mogztter) [inline lib/converter.js]: It should be `/document.css` to be consistent with `<link href="document.css" rel="stylesheet">` right ?

Comment (oncletom) [reply lib/converter.js]: Yes.

I was wary of naming collisions at this point and never completed the thinking around it.

Also, it made me think of the `stylesheet` and `linkcss` well known Asciidoctor attributes.

I'm happy to change either of these in a way which is consistent with Asciidoctor behaviour.

**Comment by Mogztter** [comment]
I'm still not sure about the embedded HTTP server. It has a few advantages but now that Paged.js is using `XMLHttpRequest` it's not mandatory.
Thoughts ?

**Comment by oncletom** [comment]
> I'm still not sure about the embedded HTTP server. It has a few advantages but now that Paged.js is using XMLHttpRequest it's not mandatory.

Agreed, it's not mandatory.

I see one use case it can be useful: when `--watch` (or an hypothetic `--preview` option) is enabled. You get a URL to check the document output as a _prepring proofreading_, with a chance to see what's wrong in the browser devtools.

Another one is if a document has a need for CORS resources. An HTTP server will be required.

At the moment this is achieved by peaking at the temporary HTML file. If that's more handy, then we can drop the internal HTTP server and favour the file for now.

**Comment by Mogztter** [comment]
> I see one use case it can be useful: when --watch (or an hypothetic --preview option) is enabled. You get a URL to check the document output as a prepring proofreading, with a chance to see what's wrong in the browser devtools.
> Another one is if a document has a need for CORS resources. An HTTP server will be required.
> At the moment this is achieved by peaking at the temporary HTML file. If that's more handy, then we can drop the internal HTTP server and favour the file for now.

Using an internal HTTP server definitely makes sense and in the end we might introduce one but I don't want to commit to it just yet.
The reason is that I want to make sure that the converter design is good enough. I'm still not sure about the separation of concerns when it comes to editing/designing a PDF. I'm also still wondering about how we should bundle things together to make sure that the ressources are properly resolved.

**Comment by oncletom** [comment]
Agreed 👍

**Pull request closed without merging**